### ISSUE TITLE
support xorg-x11-drv-libinput

### ISSUE THREAD

**Issue by raveit65** [mate-desktop/mate-control-center#133]
There is a new change proposal for fedora 22.

```
On Fri, Dec 12, 2014 at 03:38:04PM +0100, Rave it wrote:
> Am Fri, 12 Dec 2014 12:00:07 +0000
> schrieb devel-announce-request@lists.fedoraproject.org:
>   
> > Message: 1
> > Date: Thu, 11 Dec 2014 14:42:11 +0100
> > From: Jaroslav Reznik <jreznik@redhat.com>
> > To: devel-announce@lists.fedoraproject.org
> > Subject: F22 System Wide Change: Change xorg input stack to use
> >     libinput
> > Message-ID: <4092034.fSnsvv0Gs3@dhcp-0-163.brq.redhat.com>
> > Content-Type: text/plain; charset="us-ascii"
> > 
> > = Proposed System Wide Change: Change xorg input stack to use libinput =
> > https://fedoraproject.org/wiki/Changes/LibinputForXorg
> > 
> > Change owner(s): Hans de Goede <hdegoede@redhat.com>
> > 
> > Replace the current (low-level) input xorg drivers with libinput using the 
> > xorg-x11-drv-libinput wrapper. 
> > 
> > == Detailed Description ==
> > Currently xorg uses different input drivers depending on the device type. This 
> > makes it impossible to do things like middle button scrolling on the 
> > trackpoint on laptops where the trackpoint buttons are software-emulated 
> > buttons on the touchpad. Besides this the xf86-input-synaptics driver was 
> > never really designed for multi-touch touchpads and this causes various 
> > issues.
> > 
> > For Wayland we've been working on a new improved input stack, which is to be 
> > shared by all compositors and lives inside libinput. We plan to replace the 
> > current (low-level) input xorg drivers with libinput using the xorg-x11-drv-
> > libinput wrapper. 
> > 
> > == Scope ==
> > Besides xorg changes, this will also require changes to the control panel 
> > applets for mouse / touchpad configuration in the various desktop environments, 
> > as those all are hardcoded to use the xorg-x11-drv-synaptics specific 
> > interfaces.
> > 
> > * Proposal owners:
> > Package libinput and xorg-drv-input-libinput (done), make sure that xorg-drv-
> > input-libinput has the necessary config interfaces for control panel 
> > mouse/touchpad config applets (wip). Write patches for gnome-control-center 
> > mouse/touchpad capplet. Coordinate with other desktop environments.
> > 
> > * Other developers:
> > GNOME: merge the gnome-control-center patches. KDE: limits itself to standard 
> > X11 mouse config interfaces, no changes needed. Other Desktop Environments: 
> > adjust control-panel code to deal with xorg-x11-drv-libinput, merge these 
> > changes.
> > 
> > * Release engineering: N/A
> > * Policies and guidelines: N/A  
> 
> I would be very helpful if you could target a x-server version when
> control-center apps should be ready for this change, to help upstreams.
> Can we expect that those x-server changes also land in other distro later?
> Or is this limited to fedora only?  

The X server version is independent of this change. You can install
and use xorg-x11-drv-libinput right now and while I haven't tested it, the
driver probably works with anything newer than F18 or so.

Cheers,
   Peter
```
##

**Comment by NiceandGently** [comment]
the libinput patches itself i found here
http://lists.freedesktop.org/archives/wayland-devel/2014-April/014311.html
but i found nothing for control-center.

**Comment by NiceandGently** [comment]
I found some commits in gnome-settings-daemon.
https://git.gnome.org/browse/gnome-settings-daemon/log/?h=wip/settings-relocation
and it looks like that those ones is for control-center.
https://git.gnome.org/browse/gnome-control-center/log/?qt=grep&q=mouse
https://git.gnome.org/browse/gnome-settings-daemon/log/?h=wip%2Fsettings-relocation&qt=grep&q=wacom

**Comment by alxpl** [comment]
Hi, I have just filed
https://bugzilla.redhat.com/show_bug.cgi?id=1225629

I don't know if the missing touchpad tab has anything to do with libinput, but my settings not taking effect, probably has.

Let me know if you need any more information.

**Comment by NiceandGently** [comment]
@alxpl 
Does the issue exists if you **don't use** gdm as display-manager.
Please try out lightdm.

**Comment by alxpl** [comment]
Do you prefer to continue the conversation here or at rhbz?

**Comment by NiceandGently** [comment]
The issue here is that libinput driver is prefered in xorg configuration with fedora 22.
```
[rave@satellite ~]$ ls /usr/share/X11/xorg.conf.d/
10-evdev.conf   50-synaptics.conf  50-wacom.conf
10-quirks.conf  50-vmmouse.conf    90-libinput.conf
```
I could fixed that with shiping a 99-synaptics-mate.conf via mate-desktop package.
http://pkgs.fedoraproject.org/cgit/mate-desktop.git/commit/
This will not break the gnome-shell configuration tool , i tested it.
But it would be nice if mate-mouse-properties will support libinput driver before next fedora release.
It is posible that synaptics driver will removed for next release, and this is definetly a breakage for mate desktop in fedora.

**Comment by raveit65** [comment]
https://bugzilla.redhat.com/show_bug.cgi?id=1338585
Looks like that gnome have dropped support with gnome-3.20 in f24.
So, it isn't possible anymore to install both desktops on the same system.
In other words mate isn't compatible anymore to gnome or vice versa.
As libinput comes from Xorg and not from gnome it is our problem and not a gnome problem.
@monsta @flexiondotorg @clefebvre 
We need support for libinput in the foreseeable future.

**Comment by monsta** [comment]
Damn... and I'll have to keep synaptics support for LMDE 2 which doesn't have Xorg's libinput package yet. :confused:

**Comment by raveit65** [comment]
well, it shouldn't be a prob to support both, see my link.
2016-01-04	mouse: Remove support for non-libinput mouse configurations
This is the culprit commit in gnome-3.20, before this gnome did support both drivers.

**Comment by raveit65** [comment]
It looks like that Xorg have forced gnome to remove support for synaptic driver.
https://cgit.freedesktop.org/xorg/driver/xf86-input-synaptics/commit/?id=59e5db025307404fbfbc82f2fb3fe91d6a3005d7

**Comment by whot** [comment]
> It looks like that Xorg have forced gnome to remove support for synaptic driver.

I'm not sure how that commit can be interpreted that way. The synaptics driver is already overridden by libinput on every install, the commit above simply removes the need for the symlink dance. If synaptics is installed, it will be used. If it isn't installed, it won't be used and we fall back to libinput.

**Comment by raveit65** [comment]
Yes, but until synaptic driver isn't removed from default workstation installation in fedora, using a priority than libinput driver is a problem for gnome.
As far i can see the comps report is open.
Beside from this it is imposible to use Mate and gnome on a same system now in f24.

**Comment by willysr** [comment]
libinput is not going to be added in Slackware soon (at least not in 14.2), so i was hoping if it should be considered optional and NOT a mandatory package for MATE

**Comment by whot** [comment]
fwiw, I've had a look at adding libinput support to mate-settings-daemon today and it's going to be a fairly time-consuming task due to the amount of cleanup work required to hook the bits in. And some functionality will be lost because libinput doesn't allow for some configuration settings that you currently expose. Do you have a branch currently tracking the libinput work? I can probably find the odd spare cycle to contribute, but I won't write the whole mate support.

**Comment by monsta** [comment]
No, we still didn't start working on this. The team is too small and busy with other issues, not to mention the amount of work on GTK+3 support... We'd appreciate some help with this.

I'd like to keep synaptics supported (for distros that have it) while adding libinput support.

**Comment by whot** [comment]
what's the gist on the GTK+3 support - will settings-daemon be ported to gtk3?

**Comment by monsta** [comment]
Eventually yes, but not right now. Just like other apps, it has GTK+3 support, you can build it with it. It's just that GTK+2 is the default one.

**Comment by whot** [comment]
right. main reason I'm asking is that cinnamon was relatively simple because they already have a lot of the code cleanups that the gnome guys did for gnome3. those are missing in mate and it makes the task rather daunting, because a lot of code has to be shuffled around and tested before we can add the libinput hooks...

**Comment by monsta** [comment]
We backport fixes and cleanups from gnome apps from time to time, so we can get on it. We're near 1.16 release already though, so all work will probably go in the next dev phase...

**Comment by monsta** [comment]
@whot: https://github.com/linuxmint/cinnamon-settings-daemon/pull/131 seems to be fairly easy to port to m-s-d. Or do we need to pick some more commits from gnome? I might miss something.

**Comment by monsta** [comment]
Forgot to tell that we're moving to GTK+3 only code for 1.18, so current mess of ```#if```s won't be a problem :smile:

**Comment by whot** [comment]
@monsta  yes, it's fairly easy to port in terms of implementation, you pretty much do exactly the same thing. But the problem is that mate is lacking a whole bunch of general cleanup that cinnamon got before forking, the code is more spaghetti-like and you can't just slot the new bits in as easily. so the large part of the mate work would be to clean bits up, factor them out into helpers, etc. so that something like the patch linked to can go on top.

**Comment by monsta** [comment]
Aha... I thought the cleanup done in some commits of https://github.com/linuxmint/cinnamon-settings-daemon/pull/131 would be enough.

**Comment by monsta** [comment]
Ok, found some differences and some pieces apparently not ported from gnome... will try to fix that.

**Comment by monsta** [comment]
Almost there... had to backport some new touchpad settings as well.
Since we'll have to modify both m-s-d and m-c-c, I've decided to create a separate issue for discussing m-s-d changes: https://github.com/mate-desktop/mate-settings-daemon/issues/160.

**Comment by monsta** [comment]
Ok, m-s-d has libinput support now. Now on to UI changes...

I've made ```dev-libinput``` branch here for testing m-c-c with libinput. So far, it should detect libinput touchpads and enable the touchpad settings tab.

All UI elements are enabled now, we need to think what and how to disable when libinput is used instead of Synaptics...

**Comment by monsta** [comment]
@flexiondotorg @raveit65 libinput commits rebased and merged to master, https://github.com/mate-desktop/mate-control-center/commit/678aa95188f518c05169084d04aa31e49035ee6b and https://github.com/mate-desktop/mate-control-center/commit/9d6f50fb73baabce06063e6f377aefd1e82a4f7d.
I'll leave this open for reasons stated above...

**Comment by raveit65** [comment]
> All UI elements are enabled now, we need to think what and how to disable when libinput is used instead of Synaptics...

@monsta 
I recall during testing that we need to revert some commits from clem for using libinput?
If yes, tell me which ones. I will try to create patches for fedora and ubuntu.

**Comment by clefebvre** [comment]
Hi @raveit65 if I remember well, I made it possible to enable/disable edge and two-finger scrolling independently, and that isn't possible with libinput.

**Comment by clefebvre** [comment]
https://github.com/mate-desktop/mate-control-center/commit/f6f179c4ad0494aa0fa65ea9c2ba68fd99713a4d

**Comment by raveit65** [comment]
Hi @clefebvre 
I guess this one from m-s-d needs revert too ?
https://github.com/mate-desktop/mate-settings-daemon/commit/02ef9ca6f5ac17e9c68d1cafa5f2581fa824dee8

**Comment by raveit65** [comment]
Imo best way to handle that should be a configure option to use it with libinput or synaptics.

**Comment by whot** [comment]
do you actually have a use-case for enabling edge *and* two-finger scrolling at the same time?
likewise with the separate horiz scrolling - do you have a use-case where someone has e.g. horiz edge scrolling but vert two-finger scrolling? because if not, having a single "horiz enabled" toggle is a lot easier to handle...

**Comment by monsta** [comment]
@whot: I'm not even sure there was a demand for that feature... maybe @clefebvre remembers more than me

**Comment by monsta** [comment]
I think we'll have to redesign some parts of UI and possibly make some new m-s-d settings to better handle both libinput and synaptics. Besides the mentioned scrolling settings thing, there's also an issue with clickfinger setting (https://github.com/mate-desktop/mate-settings-daemon/pull/183 and https://github.com/mate-desktop/mate-control-center/pull/289). This is getting outside of scope of just one bug report, so I'll probably create a team discussion for that. Anyway, it's a work for 1.22, as we're close to releasing 1.20 now.
